tools/cryptogenotyper/cryptogenotyper.xml

```diff
@@ -0,0 +1,141 @@
+<tool id="CryptoGenotyper" name = "CryptoGenotyper" version="@VERSION@">
+  <description>
+    CryptoGenotyper is a standalone tool to analyze Cryptosporidium Sanger sequencing data and classify the species and subtype based on SSU rRNA and gp60 gene markers, respectively.
+  </description>
+  <macros>
+   <token name="@VERSION@">1.0</token>
+  </macros>
+  <requirements>
+    <requirement type="package" version ="@VERSION@">cryptogenotyper</requirement>
+  </requirements>
+  <version_command>cryptogenotyper --version</version_command>
+  <command detect_errors="exit_code">
+     <![CDATA[
+
+     #set $ref_file=''
+
+     #if $reference.ref == "no"
+        ln -s "${reference.db}" "${reference.db.name}" &&
+        #set $ref_file = $reference.db.name
+     #end if
+
+     #if $primers["seqtype"] == "contig"
+        ln -s "${$primers.abi_input["forward"]}" "${primers.abi_input.name}_forward.ab1" &&
+        ln -s "${$primers.abi_input["reverse"]}" "${primers.abi_input.name}_reverse.ab1" &&
+        cryptogenotyper -i "." -m "$marker" -t "$primers.seqtype" -f "forward" -r "reverse"
+        #if $reference.ref == "no"
+          --databasefile $ref_file
+        #end if
+     #else
+        ln -s "${primers.abi_input}" "${primers.abi_input.element_identifier}" &&
+        cryptogenotyper -i "./${primers.abi_input.element_identifier}" -m "$marker" -t "$primers.seqtype" -f "" -r ""
+        #if $reference.ref == "no"
+          --databasefile $ref_file
+        #end if
+     #end if
+     #if $outputheader == "no"
+           --noheaderline
+     #end if
+     -o "result";
+     ]]>
+  </command>
+  <inputs>
+      <param name="marker" type="select" label="Marker">
+          <option value="18S">SSU rRNA</option>
+          <option value="gp60">gp60</option>
+      </param>
+      <conditional name="reference">
+        <param name="ref" type="select" label="Use default reference file?">
+            <option value="yes">Yes</option>
+            <option value="no">No</option>
+        </param>
+        <when value="yes">
+        </when>
+        <when value="no">
+            <param name="db" type="data" format="fasta" label="Reference Database File:" help=".fa fasta file type"/>
```

Comment:
you could remove the entire conditional and make this parameter optional="true"

Reply:
I had tried with the optional parameter option, but then I need `<param name="db" type="data" format="fasta" label="Reference Database File:" help=".fa fasta file type"/>` to be executed when user selects No (meaning that user wants to user his/her own custom database of reference alleles). I know it is a bit ugly from the code perspective, but I do not know how to make it work when decorative "Yes" option is selected as a filler. Thus I left it as it is.

Reply:
Sorry, that I do not get. Can you not remove the entire condition and just put `<param name="db" type="data" optional="true" format="fasta" label="Reference Database File:" help=".fa fasta file type"/>`?

I'm not worried about the code so much, but the UX should be better because you need on less click :)

Reply:
I've tried to make it optional but then it crashes as then it is possible to launch the program when there is no custom database specified. Otherwise if will make it non-optional, user will see a red section highlighted preventing the launch. By default user does not need to specify a database as the default one is used instead.
Maybe we can make it such that this custom field is always visible but is optional so there is no need to select Yes or No. If it is defined, then it is used.

Reply:
Oh I think I understand now. What you need to do is add something like this to cheetah:

```
#if $db:
   do something when a file was selected
#end if
```

Reply:
Yes, let's make it always visible field and then if it is defined, execute that parameter. There will be 1 less click in case user would decide to provide a custom DB. We are now releasing a version +galaxy 0 with all changes. Many thanks!